Sorting a LinkedList: does not deal with firstLink and lastLink if th…

dupriezt opened the pull request:
Fixing issue: https://pharo.fogbugz.com/f/cases/22571/Sorting-empty-LinkedList-and-Interval-raises-an-exception.

Sorting a LinkedList: does not deal with firstLink and lastLink if the list is empty to avoid an exception when sorting an empty LinkedList.

Sorting an Interval: directly returns the Interval, since sorting an Interval does not make sense anyway.

MarcusDenker committed:
22569 add two more inspector views for MetaLink: nodes and methods
https://pharo.fogbugz.com/f/cases/22569/add-two-more-inspector-views-for-MetaLink-nodes-and-methods

Commit:
Added getter and setter methods to Reflectivity class for the argsAsArray option. With comments about what this option does in the setter method.

MarcusDenker committed:
22570  add #method reification
https://pharo.fogbugz.com/f/cases/22570/add-method-reification

Commit:
Sorting a LinkedList: does not deal with firstLink and lastLink if the list is empty to avoid an exception when sorting an empty LinkedList.
Sorting an Interval: directly returns the Interval, since sorting an Interval does not make sense anyway.

Commit:
Sorting a LinkedList: does not deal with firstLink and lastLink if the list is empty to avoid an exception when sorting an empty LinkedList.
Sorting an Interval: directly returns the Interval, since sorting an Interval does not make sense anyway.

MarcusDenker committed:
Merge pull request #1905 from MarcusDenker/22569-add-two-more-inspector-views-for-MetaLink-nodes-and-methods

22569-add-two-more-inspector-views-for-MetaLink-nodes-and-methods

MarcusDenker committed:
Merge pull request #1906 from dupriezt/Reflectivity-optionArgsAsArray-method

Added getter and setter methods to Reflectivity class for the argsAsA…

astares committed:
22578 Tag ManifestCollectionsTests

https://pharo.fogbugz.com/f/cases/22578/

astares committed:
22579 Tag and comment ManifestSystemHashing

https://pharo.fogbugz.com/f/cases/22579/Tag-and-comment-ManifestSystemHashing

juliendelplanque commented on src/Collections-Sequenceable/LinkedList.class.st:
```diff
@@ -549,8 +549,10 @@ LinkedList >> size [
 LinkedList >> sort: aBlock [
 	|aLinkedList|
 	aLinkedList := LinkedList newFrom: (self asArray sort: aBlock).
```
If the `LinkedList` is empty, does it really make sense to execute this line?

Maybe `self isEmpty` should be a guard clause for the whole method which returns a new (empty) linked list without doing the conversion to array and sorting stuff.

dupriezt replied:
Makes sense. Oversight on my end.
I made a new issue and a new branch for the LinkedList change: #1921

juliendelplanque commented on src/STON-Core/Interval.extension.st:
```diff
@@ -12,6 +12,12 @@ Interval >> fromSton: stonReader [
 	stonReader parseNamedInstVarsFor: self
 ]
 
+{ #category : #'*STON-Core' }
```
I think this change should not be in this PR.

1. This is a different issue.

2. Is it an extension method of `STON-Core` package on purpose? If it is, it sustain even more 1.
If it is not, the method should be put in the right package.

dupriezt replied:
No idea why the method I wrote turned out to be an extension of STON-Core.

dupriezt replied:
I made a new issue, branch and pull request for this: https://github.com/pharo-project/pharo/pull/1922
It's just adding the Interval>>sort: method and having it raise an "Can not sort an Interval" error.

juliendelplanque replied:
Why not just *not* implement `#sort:` to get a DNU?

dupriezt replied:
-> https://github.com/pharo-project/pharo/pull/1908#discussion_r225918526 :)

juliendelplanque commented on src/STON-Core/Interval.extension.st:
```diff
@@ -12,6 +12,12 @@ Interval >> fromSton: stonReader [
 
+{ #category : #'*STON-Core' }
+Interval >> sort: aBlock [
+	"Directly returns the Interval, since sorting an Interval does not make sense."
```
I think we should not say it does not make sense to `#sort:` an `Interval` in the comment. Else, the method should simply not be implemented.

It may make sense that one can call `#sort:` on an interval but I'm not sure how it should be implemented.

For example, if the following evaluate correctly, it might be ok:
```
((1 to: 10) sorted: [ :a :b | a > b ]) >>> (10 to: 1 by: -1)
```

But in some case, depending on the sort block, an instance of `Interval` can not be returned.

dupriezt replied:
Interval is a subclass of SequenceableCollection which has a default implementation of #sort:
This default implementation results in an exception when used on an Interval because you cannot #at:put: in an Interval.
"does not make sense" is maybe a bit too much. I meant by that that an Interval is not a concept that has an intuitive notion of what it means to sort it.
So I implemented a trivial sort function to override the default one from SequenceableCollection to at least not raise a strange exception when someone sorts an Interval by accident.
I guess that the sort method of Interval could raise an explicit "Cannot sort an Interval" exception instead of the "you can not store into an interval" exception raised by Interval>>at:put:

dupriezt replied:
I made a new issue, branch and pull request for this: https://github.com/pharo-project/pharo/pull/1922
It's just adding the Interval>>sort: method and having it raise an "Can not sort an Interval" error.

juliendelplanque replied:
Ah, ok nervermind may comment on the other thread, I didn't know Sequenceable collection implement `#sort:`.

juliendelplanque reviewed (changes requested):
As a summary, I think the code in `LinkedList` should be slightly enhanced before integration and that the code in `Interval` should be reviewed by someone else to discuss if it is right or not.

But it should be in another issue anyway.

MarcusDenker committed:
Merge pull request #1911 from astares/22579-Tag-and-comment-ManifestSystemHashing

22579 Tag and comment ManifestSystemHashing

MarcusDenker committed:
Merge pull request #1910 from astares/22578-Tag-ManifestCollectionsTests

22578 Tag ManifestCollectionsTests

astares committed:
22580 Tag ManifestCollectionsUnordered and related classes in package Collections-Unordered

https://pharo.fogbugz.com/f/cases/22580

astares committed:
22581 Tag ManifestCollectionsStrings and related classes in package Collections-Strings

https://pharo.fogbugz.com/f/cases/22581/Tag-ManifestCollectionsStrings-and-related-classes-in-package-Collections-Strings

MarcusDenker committed:
Revert "22564-Spec-tests-has-a-class-that-need-to-be-moved-to-deprecated-pacakage"

astares committed:
22582 Tag ManifestSystemFinalization and classes in System-Finalization

astares committed:
22583 Tag ManifestCollectionsStreams and related classes in Collections-Streams

https://pharo.fogbugz.com/f/cases/22583

astares committed:
22584 Tag ManifestCollectionsSupport and related classes in Collections-Support

https://pharo.fogbugz.com/f/cases/22584/

MarcusDenker committed:
Merge pull request #1914 from pharo-project/revert-1900-22564-Spec-tests-has-a-class-that-need-to-be-moved-to-deprecated-pacakage

Revert "22564-Spec-tests-has-a-class-that-need-to-be-moved-to-deprecated-pacakage"

MarcusDenker committed:
Merge pull request #1907 from MarcusDenker/22570-add-method-reification-

22570-add-method-reification-

MarcusDenker committed:
Merge pull request #1915 from astares/22582-Tag-ManifestSystemFinalization-and-classes-in-System-Finalization

22582 Tag ManifestSystemFinalization and classes in System-Finalization

MarcusDenker committed:
Merge pull request #1913 from astares/22581-Tag-ManifestCollectionsStrings-and-related-classes-in-package-Collections-Strings

22581 Tag ManifestCollectionsStrings and related classes in package Collections-Strings

MarcusDenker committed:
Merge pull request #1912 from astares/22580-Tag-ManifestCollectionsUnordered-and-related-classes-in-package-Collections-Unordered

22580 Tag ManifestCollectionsUnordered and related classes in package Collections-Unordered

MarcusDenker committed:
Merge pull request #1917 from astares/22584-Tag-ManifestCollectionsSupport-and-related-classes-in-Collections-Support

22584 Tag ManifestCollectionsSupport and related classes in Collections-Support

MarcusDenker committed:
Merge pull request #1916 from astares/22583-Tag-ManifestCollectionsStreams-and-related-classes-in-Collections-Streams

22583 Tag ManifestCollectionsStreams and related classes in Collections-Streams

jecisc committed:
Remove some unreferenced Spec classes

Fixes https://pharo.fogbugz.com/f/cases/22564/Spec-tests-has-a-class-that-need-to-be-moved-to-deprecated-pacakage

MarcusDenker committed:
Merge pull request #1918 from jecisc/22564-Spec-tests-has-a-class-that-need-to-be-moved-to-deprecated-pacakage

22564-Spec-tests-has-a-class-that-need-to-be-moved-to-deprecated-pacakage

MarcusDenker commented:
this PR was done against a version were the bootstrap was broken. As the PR check uses this version, the PR test check stays broken even though we fixed the broken build in the meantime.

Solution: update the PR to be against the latest version.

Commit:
Merge branch '22571-Sorting-empty-LinkedList-and-Interval-raises-an-exception' of github.com:dupriezt/pharo into 22571-Sorting-empty-LinkedList-and-Interval-raises-an-exception

Commit:
Refactored sort: method of LinkedList to optimise the case where the LinkedList is empty.

dupriezt commented:
Improved LinkedList>>sort: change -> new pull request  #1921
Interval>>sort: change -> new pull request #1922
